tao_compiler/mlir/disc/transforms/disc_dot_merge.cc

```diff
@@ -190,6 +236,361 @@ void BuildDotClusters(Block* block,
+      } 
+    }
+    auto lhs_type =
+        RankedTensorType::get(lhs_shapes, orig_lhs_type.getElementType());
+    lhs = builder.create<mhlo::ConcatenateOp>(loc, lhs_type, lhs_operands,
+                                                    concat_dim);
+    rhs = foremost_dot.rhs();
+  } else {
+    assert("share_type error");
+  }
+  
+  // Concat Result type.
+  auto result_rank = orig_result_type.getRank();
+  SmallVector<int64_t, 4> result_shapes(result_rank, ShapedType::kDynamicSize);
+  for (int64_t i = 0; i < result_rank; i++) {
+    if (i == concat_dim) {
+      if (is_dynamic_shape) {
+        result_shapes[i] = -1;
+      } else {
+        result_shapes[i] = concat_dim_sum;
+      }
+    } else {
+      result_shapes[i] = orig_result_type.getDimSize(i);  
+    }   
+  }
+  auto result_type =
+      RankedTensorType::get(result_shapes, orig_result_type.getElementType());
+  // Build dot dimension numbers.
+  auto dim_numbers = foremost_dot.dot_dimension_numbers();
+
+  SmallVector<int64_t> lhs_batching_dims;
+  auto lhs_batch = dim_numbers.getLhsBatchingDimensions();
+  lhs_batching_dims.insert(lhs_batching_dims.end(), lhs_batch.begin(),
+                           lhs_batch.end());
+
+  SmallVector<int64_t> rhs_batching_dims;
+  auto rhs_batch = dim_numbers.getRhsBatchingDimensions();
+  rhs_batching_dims.insert(rhs_batching_dims.end(), rhs_batch.begin(),
+                           rhs_batch.end());
+
+  SmallVector<int64_t> lhs_contracting_dims;
+  auto lhs_contract = dim_numbers.getLhsContractingDimensions();
+  for (auto& val : lhs_contract) {
+    lhs_contracting_dims.push_back(val);
+  }
+
+  SmallVector<int64_t> rhs_contracting_dims;
+  auto rhs_contract = dim_numbers.getRhsContractingDimensions();
+  for (auto& val : rhs_contract) {
+    rhs_contracting_dims.push_back(val);
+  }
+
+  auto dot_dimension_attr = mhlo::DotDimensionNumbersAttr::get(
+      builder.getContext(), lhs_batching_dims, rhs_batching_dims,
+      lhs_contracting_dims, rhs_contracting_dims);
+  
+  // Build new DotGeneralOp
+  Value merged_dot = builder.create<mhlo::DotGeneralOp>(
+      loc, result_type, lhs, rhs, dot_dimension_attr, nullptr);
+  // Build slice  for each of the original dot op, and replace the dot.  
+
+  if (!is_dynamic_shape) {
+    // Use static-dim ops.
+    int64_t concat_dim_start = 0;
+    for (int64_t i = 0; i < ops.size(); i++) {
+      mhlo::DotGeneralOp op = dyn_cast<mhlo::DotGeneralOp>(ops[i]);
+      auto orig_dot_type = op.getType().dyn_cast<RankedTensorType>();
+
+      SmallVector<int64_t> start(result_rank);
+      SmallVector<int64_t> limit(result_rank);   
+      SmallVector<int64_t> strides(result_rank, 1);
+      
+      for (int64_t j = 0; j < result_rank; j++) {
+        if (j == concat_dim) {
+          start[j] = concat_dim_start;
+          limit[j] = concat_dim_start +
+                    orig_dot_type.getDimSize(concat_dim);
+          concat_dim_start = limit[j];
+        } else {
+          start[j] = 0;
```

Comment:
Better the specify the init value of `start` as `0` when creating this vector. And only initialize `start[concat_dim]` outside the loop.